### ISSUE TITLE
Make ExampleCrossModRecipe

### ISSUE THREAD

**Issue by Setnour6** [bluemagic123/tModLoader#2070]
### Do you intend to personally contribute/program this feature?

No

### Description

So I have been trying to work on one of my modifications to add recipes for an item belonging to a different modification, such as the ExampleMod.

So we know that we can make recipes with vanilla items and also for items that your/our/my mod has:
`Recipe recipe = Mod.CreateRecipe(ModContent.ItemType<Items.ExampleItem>(), 999);`

But we don't have the ability to add modded items to use in recipes, like back in 1.3 when we use this:
`recipe.AddIngredient((ModLoader.GetMod("ThoriumMod").ItemType("ThoriumItemGoesHere")), 10);`

How can I make recipes that use modded items in tModLoader 1.4 Alpha? Please let me know if there is a way to do so.
Once again, please let me know.

### What does this proposal attempt to solve or improve?

So basically, this feature intends to make it so anybody can add recipe ingredients from other mods like back in 1.3, but is applied to 1.4 Alpha.

`Recipe recipe = Mod.CreateRecipe(ModContent.ItemType<Items.ExampleItem>(), 999);` but it uses a different mod's item(s).

**Comment by Solxanich** [comment]
This probably should have been asked first in #alpha-modding-help  in the Discord.

You can use something like ModContent.TryFind<ModItem>("ExampleMod/ExampleItem") to get the reference to the item.

You should be able to find this info on the migration guide as well.

**Comment by Setnour6** [comment]
> This probably should have been asked first in #alpha-modding-help in the Discord.
> 
> You can use something like ModContent.TryFind("ExampleMod/ExampleItem") to get the reference to the item.
> 
> You should be able to find this info on the migration guide as well.

Unfortunately, I do not have discord. Is there any other place I can send this issue to?

I tried the code you sent here, but it didn't work.
![image](https://user-images.githubusercontent.com/70764728/147982141-f9a3d41c-c4fd-491c-8e09-dbc65ecc1992.png)
Here is the code that I've been trying to change to make it work in 1.4 Alpha:
![image](https://user-images.githubusercontent.com/70764728/147982580-f4c8f39e-542f-460e-8566-f802e46a25e5.png)
Note: The main errors to solve are the "TryGetMod" and "TryFind". Everything else with a red line is on purpose.

I have looked through the migration guide, and nothing worked. The mod is freshly built from 1.4 Alpha.

Is there any other solution, or maybe I'm missing something? Please let me know.

**Comment by Chicken-Bones** [comment]
Your best option is:
```cs
if (!ModContent.TryFind("ExampleMod/ExampleItem", out var exampleItem) return;
CreateRecipe(exampleItem)
...
```

You can also throw an exception instead of returning.

You can put all the `ModContent.TryFind("Mod/ItemName", out var itemVariable)` calls you need at the top of your recipes method, and each `out` variable will be filled with the corresponding `ModItem`

**Comment by Setnour6** [comment]
> Your best option is:
> 
> ```cs
> if (!ModContent.TryFind("ExampleMod/ExampleItem", out var exampleItem) return;
> CreateRecipe(exampleItem)
> ...
> ```
> 
> You can also throw an exception instead of returning.
> 
> You can put all the `ModContent.TryFind("Mod/ItemName", out var itemVariable)` calls you need at the top of your recipes method, and each `out` variable will be filled with the corresponding `ModItem`

I don't even know if I fully understand the assignment, but I tried to do what you say anyways, and tried once again to follow the instructed from the Migration guide in case I am missing something, but still errors occured.
![image](https://user-images.githubusercontent.com/70764728/147994255-5e6c8f1c-7690-4aa5-bd65-4b98c4730ab9.png)

Here is all of the "using" calls (I forgot what they are called) that are being used:
![image](https://user-images.githubusercontent.com/70764728/147994328-1ff137a4-5e0e-4a21-a219-c8c5f69b8807.png)
The target framework is net6.0. I'm using Visual Studio 2019. I think it could be the way VS2019 is reading the information, or the framework (netcore wouldn't really work, since I'd be told to upgrade csproj on the mod sources).

**Comment by Solxanich** [comment]
Just FYI - VS2019 doesn't fully support .net6. That would be a possible reason for the type inference failing.

With respect to your issue, Try adding ``<ModItem>``  after TryFind so its TryFind ``<ModItem>`` () 
Its expecting a type declaration in TryFind. 

I'd also consider upgrading to VS2022

**Comment by Chicken-Bones** [comment]
Sorry, change `out var` to `out ModItem`

**Comment by Setnour6** [comment]
So I upgraded to VS2022, followed everything that was said once again in how I perceive it, and I still receive the same errors in the methods.
![image](https://user-images.githubusercontent.com/70764728/147999276-81b7dcf1-1f0e-4f66-9c14-bba5cc8259cc.png)
_Also applies when adding `()`_

![image](https://user-images.githubusercontent.com/70764728/147999292-ccc928da-98eb-43dc-902c-b73e1c70e559.png)
_Also applies with `TryFindMod`_

I could be doing this wrong:
![image](https://user-images.githubusercontent.com/70764728/147998925-9723df09-597a-4ebc-b017-082505fc3deb.png)
![image](https://user-images.githubusercontent.com/70764728/147998985-04124214-8432-4f7e-b7b1-5158e1dc93f3.png)
_Even without `CreateRecipe` listed here, error persists. Also wrote `TryGetMod`/`TryFindMod` but still there were errors_

![image](https://user-images.githubusercontent.com/70764728/147999099-b8e553b9-3de8-4de5-9393-53b53bab3102.png)
![image](https://user-images.githubusercontent.com/70764728/147999177-e6bae44d-71bd-4b0c-9fe3-932d84363edb.png)
_Errors still persist without `CreateRecipes` usage in the code above, and use of `TryGetMod`/`TryFindMod`_.

The entire recipe code:
![image](https://user-images.githubusercontent.com/70764728/147999507-a055a17e-2195-4d76-aa45-fdeca1091824.png)
_I also wrote `TryGetMod`/`TryFindMod`, but still, errors persisted. (It says "TryMod" by accident. Still... errors...)_

Here is the code section:
```cs
public class YourModName : Mod
	{
	ModContent.TryFind("Mod/ItemName", out var itemVariable)
		public override void AddRecipes()
		{

			if (!ModContent.TryFind("ExampleMod/ExampleItem", out var exampleItem) return; //Also wrote TryFind(), TryFindMod and TryGetMod. Error persisted.
				CreateRecipe(exampleItem)
			.AddIngredient<Items.YourModItem>()
			.AddRecipeGroup("Mod:ItemGroup")
			.AddTile(TileID.VanillaTileID)
			.Register();

		}
	}
```

I'm very sure that I am still making mistakes with the code and everything.

**Since `out var` to `out ModItem`**
I tried that and...
![image](https://user-images.githubusercontent.com/70764728/148001658-7d83f6f0-cf90-4612-b96b-7ca4e84604d7.png)
![image](https://user-images.githubusercontent.com/70764728/148001663-c7ec62c3-026f-4f5f-a1f3-1c72dab99eb2.png)

![image](https://user-images.githubusercontent.com/70764728/148001818-5e820682-fc27-4bda-8dda-ffd2fc38b95e.png)
![image](https://user-images.githubusercontent.com/70764728/148001835-64c26406-d22f-4a65-8746-923195ff3aa0.png)

I still very sure that I'm making mistakes.

**Since other comment changes**

I ended up with this:
![image](https://user-images.githubusercontent.com/70764728/148002711-4c95327c-5287-4ddd-bd12-ddf474d7bd7b.png)
9 is just a placeholder, and nothing else works if I change 9 to any other value other than an integer. Is there anything else I have to do, or maybe I'm still doing this wrong?

**Comment by Setnour6** [comment]
I also keep getting MSB3270: There was a mismatch between the processor architecture of the project being built 'arch' and the processor architecture of the reference 'reference', 'arch'. What should I do for there, in case it conflicts with everything.

Also, I tried again and got this: 
![image](https://user-images.githubusercontent.com/70764728/148006299-c90e1621-06fe-4a74-83bd-7ff1ba8c42a0.png)
Still error. `CreateRecipe(ExampleItem)` where error is "ExampleItem"
"ExampleMod/ExampleItem" yeilds same result: error.

**Comment by Chicken-Bones** [comment]
Ah, looks like it's
```cs
if (!ModContent.TryFind("ExampleMod/ExampleItem", out ModItem exampleItem) return;
exampleItem.CreateRecipe()
...
```

**Comment by Chicken-Bones** [comment]
You should really make a free Discord account, you can use it via the web browser.

**Comment by Setnour6** [comment]
> You should really make a free Discord account, you can use it via the web browser.

I'd rather not, but you can message me via email at set.nour6@gmail.com.

**Comment by Setnour6** [comment]
It worked! I am in positive shock and I thank you very much @Solxanich and @Chicken-Bones for helping me out!
I will let you know if I have any other questions or concerns.
I guess this thread can be archived or deleted or idk what (or you can keep it).
Once again, thank you so much for helping me!

**Comment by Setnour6** [comment]
Hello again!
So I almost forgot to say this, but I want to add modded ingredients, for use with the cross mod recipe.
Remember in 1.3: `recipe.AddIngredient((ModLoader.GetMod("ThoriumMod").ItemType("ThoriumItemGoesHere")), 10);`
Now: `AddIngredient` something...
Here's the code I wrote, which is very likely incorrect for adding modded ingredients for CrossModRecipe (In this case, it's Qwerty's Bosses and Items):
```cs
public override void AddRecipes()
		{
			if (!ModLoader.TryGetMod("QwertyMod", out var QwertyMod))
			{
				return;
			}
			if (!ModContent.TryFind("QwertyMod/NanoDart", out ModItem NanoDart)) return; //NanoDart is used as example. This part is fine
			{
				NanoDart.CreateRecipe() //NanoDart is used as example.
			.AddIngredient(ModContent.ItemType<QwertyItemGoesHere>()) //This part contains errors
			.AddIngredient((ModContent.TryGetMod(QwertyMod)).ItemType("QwertyItemGoesHere")), 4) //This part contains errors
			.AddTile(TileID.Anvils) //This part is fine
			.Register();
			}
		}
```
What are the steps I should take when adding modded ingredients from other mods?

**Comment by steviegt6** [comment]
> Hello again! So I almost forgot to say this, but I want to add modded ingredients, for use with the cross mod recipe. Remember in 1.3: `recipe.AddIngredient((ModLoader.GetMod("ThoriumMod").ItemType("ThoriumItemGoesHere")), 10);` Now: `AddIngredient` something... Here's the code I wrote, which is very likely incorrect for adding modded ingredients for CrossModRecipe (In this case, it's Qwerty's Bosses and Items):
> 
> ```cs
> public override void AddRecipes()
> 		{
> 			if (!ModLoader.TryGetMod("QwertyMod", out var QwertyMod))
> 			{
> 				return;
> 			}
> 			if (!ModContent.TryFind("QwertyMod/NanoDart", out ModItem NanoDart)) return; //NanoDart is used as example. This part is fine
> 			{
> 				NanoDart.CreateRecipe() //NanoDart is used as example.
> 			.AddIngredient(ModContent.ItemType<QwertyItemGoesHere>()) //This part contains errors
> 			.AddIngredient((ModContent.TryGetMod(QwertyMod)).ItemType("QwertyItemGoesHere")), 4) //This part contains errors
> 			.AddTile(TileID.Anvils) //This part is fine
> 			.Register();
> 			}
> 		}
> ```
> 
> What are the steps I should take when adding modded ingredients from other mods?

> `AddIngredient(ModContent.ItemType<QwertyItemGoesHere>())`

You are not able to directly reference items from Qwerty's without an actual DLL file referenced.

> `AddIngredient((ModContent.TryGetMod(QwertyMod)).ItemType("QwertyItemGoesHere")), 4)`

You cannot do `ModContent.TryGetMod(QwertyMod)`, `QwertyMod` is already a `Mod` instance.

**Comment by Setnour6** [comment]
> > `AddIngredient(ModContent.ItemType<QwertyItemGoesHere>())`
> 
> You are not able to directly reference items from Qwerty's without an actual DLL file referenced.
> 
> > `AddIngredient((ModContent.TryGetMod(QwertyMod)).ItemType("QwertyItemGoesHere")), 4)`
> 
> You cannot do `ModContent.TryGetMod(QwertyMod)`, `QwertyMod` is already a `Mod` instance.

I'm not entirely sure what you mean by referencing a DLL file, but I added the mod dll file as a reference and wrote a using directive. There were no errors displayed in the code, but the game displayed error code CS0246: The type or namespace name 'QwertyMod' could not be found (are you missing a using directive or an assembly reference?), as well as for the item itself.

Ref (which I probably did wrong but idk):
![image](https://user-images.githubusercontent.com/70764728/148276229-aed56d3d-1c81-4799-bd96-51931cdad994.png)

Code:
`.AddIngredient(ModContent.ItemType<QwertyModItemGoesHere>()) //To reference DLL.`

Full Code:
`using QwertyMod.Content.Items.Equipment.Accessories;`
```cs
{
	NanoDart.CreateRecipe() //This already works, and still using NanoDart as an example
	.AddIngredient<Items.YourModItemGoesHere>()
	.AddIngredient<Items.QwertyModItems.YourOtherModItemGoesHere>()
	.AddIngredient(ModContent.ItemType<EnchantedWhetstone>()) //To reference DLL. using Enchanted Whetstone as an example. Error happened in-game, not csproj.
	.AddTile(TileID.DemonAltar)
	.Register();
}
```

What DLL do I actually have to reference, since the mod DLL didn't work for me? And what do I actually have to do to make my mod actually load the crossmod ingredient?

**Comment by Chicken-Bones** [comment]
The direct references approach you have is definitely the way to go if you want to build against the mod. All you need to do is to add the mod name to either `modReferences` or `weakReferences` in `build.txt`. See https://github.com/tModLoader/tModLoader/wiki/build.txt

If you want to use content from the mod without building against it (requiring the dll) then:
```cs
public override void AddRecipes()
{
	if (!ModLoader.TryGetMod("QwertyMod", out var QwertyMod)) return;
	if (!ModContent.TryFind("QwertyMod/NanoDart", out ModItem NanoDart)) return; //NanoDart is used as example. This part is fine
	if (!ModContent.TryFind("QwertyMod/QwertyItemGoesHere", out ModItem QwertyItemGoesHere)) return; // this puts a value of QwertyItemGoesHere in the variable.
	
	NanoDart.CreateRecipe() //NanoDart is used as example.
		.AddIngredient(QwertyItemGoesHere)
		.AddIngredient(QwertyItemGoesHere, 4)
		.AddTile(TileID.Anvils) //This part is fine
		.Register();
		
	// there is also the option of. In this one, you don't need to TryFind the QwertyItemGoesHere above
	NanoDart.CreateRecipe() //NanoDart is used as example.
		.AddIngredient(QwertyMod, "QwertyItemGoesHere")
		.AddIngredient(QwertyMod, "QwertyItemGoesHere", 4)
		.AddTile(TileID.Anvils) //This part is fine
		.Register();
}
```


Someone should update https://github.com/tModLoader/tModLoader/wiki/Intermediate-Recipes#cross-mod-recipes

**Comment by Setnour6** [comment]
> The direct references approach you have is definitely the way to go if you want to build against the mod. All you need to do is to add the mod name to either `modReferences` or `weakReferences` in `build.txt`. See https://github.com/tModLoader/tModLoader/wiki/build.txt
> 
> If you want to use content from the mod without building against it (requiring the dll) then:
> 
> ```cs
> public override void AddRecipes()
> {
> 	if (!ModLoader.TryGetMod("QwertyMod", out var QwertyMod)) return;
> 	if (!ModContent.TryFind("QwertyMod/NanoDart", out ModItem NanoDart)) return; //NanoDart is used as example. This part is fine
> 	if (!ModContent.TryFind("QwertyMod/QwertyItemGoesHere", out ModItem QwertyItemGoesHere)) return; // this puts a value of QwertyItemGoesHere in the variable.
> 	
> 	NanoDart.CreateRecipe() //NanoDart is used as example.
> 		.AddIngredient(QwertyItemGoesHere)
> 		.AddIngredient(QwertyItemGoesHere, 4)
> 		.AddTile(TileID.Anvils) //This part is fine
> 		.Register();
> 		
> 	// there is also the option of. In this one, you don't need to TryFind the QwertyItemGoesHere above
> 	NanoDart.CreateRecipe() //NanoDart is used as example.
> 		.AddIngredient(QwertyMod, "QwertyItemGoesHere")
> 		.AddIngredient(QwertyMod, "QwertyItemGoesHere", 4)
> 		.AddTile(TileID.Anvils) //This part is fine
> 		.Register();
> }
> ```
> 

It worked! I just added multiple recipes using Qwerty's mod, and it actually worked! Thank you very much @Steviegt6 and @Chicken-Bones for the help!
I just gave Solxan and the both of you credit in `description.txt`
![image](https://user-images.githubusercontent.com/70764728/148394821-7f50536d-9336-4c1a-b92c-29d67d53616a.png)

I'll let you all know if I have any additional questions or concerns.

> Someone should update https://github.com/tModLoader/tModLoader/wiki/Intermediate-Recipes#cross-mod-recipes

I could try to update the wiki page by adding in 1.4 information alongside 1.3 information, so everyone from either 1.3 or 1.4 alpha can follow along.
If someone other than me could update the page if I don't update the page fast enough, please do so. The information here can be used.

**Comment by Setnour6** [comment]
Hello again!

So I've been adding another mod for the cross mod recipes to add, and everything in general works, but with one added mod disabled, the recipes for the other mod is disabled.
The mods being used are Qwerty's Bosses and Items 2 and the Vitality Mod for tmodloader 1.4 alpha.

I referenced both mods under weakReferences (it was the best option for me), and also sortAfter for both mods.
![image](https://user-images.githubusercontent.com/70764728/151639950-09f5e6ab-fc07-4904-b031-000ef5791327.png)

```cs
public override void AddRecipes()
		{
			
			#region Vanilla Content
			// ALL IS WELL IN THIS SECTION
			#endregion

			#region QWERTY's Bosses and Items
			if (!ModLoader.TryGetMod("QwertyMod", out var QwertyMod)) return;
			{
				if (!ModContent.TryFind("QwertyMod/QwertyItemGoesHere", out ModItem QwertyItemGoesHere)) return; //QwertyMod

				QwertyItemGoesHere.CreateRecipe()
				.AddIngredient<Items.YourModItem>()
				.AddRecipeGroup("YourMod:YourModRecipeGroup")
				.AddIngredient(1, 5) //uses item ID 1 as example 5 times.
				.AddIngredient(QwertyMod, "QwertyIngredientItem")
				.AddTile(1) // uses tile ID 1 as example
				.Register();
				
			}
			
			#endregion

			#region Vitality Mod
			if (!ModLoader.TryGetMod("VitalityMod", out var VitalityMod)) return;
			{
				if (!ModContent.TryFind("VitalityMod/VitalityItemGoesHere", out ModItem VitalityItemGoesHere)) return; //VitalityMod

				VitalityItemGoesHere.CreateRecipe().AddIngredient<Items.YourModItem>().AddIngredient<Items.VitalityModItems.YourVitalityModItem>().AddIngredient(VitalityMod, "VialityIngredientItem").AddIngredient(1, 1).AddTile(1).AddCondition(Recipe.Condition.YourConditionGoesHere).Register(); 
				// The code works; no errors were present when building.
			}
			
			#endregion
		}
```

Once again, if one of the mods are disabled, both cross mod recipes are also removed.

Please let me know how I can have one of the mods with cross mod recipes load when one of the other mods are disabled.
Thank you.
